File: backend/omninest-media/src/main/java/com/omninest/modules/reader/domain/ReaderAnnotation.java
package com.omninest.modules.reader.domain;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.Id;
import jakarta.persistence.PrePersist;
import jakarta.persistence.PreUpdate;
import jakarta.persistence.Table;
import jakarta.persistence.Version;
import java.time.Instant;
import java.util.UUID;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

@Entity
@Table(name = "reader_annotations", schema = "omni")
@Data
@AllArgsConstructor
@NoArgsConstructor
public class ReaderAnnotation {
    @Id
    private UUID id;

    @Column(name = "owner_user_id", nullable = false)
    private UUID ownerUserId;

    @Column(name = "reader_item_id", nullable = false)
    private UUID readerItemId;

    @Column(name = "chapter_id", length = 128)
    private String chapterId;

    @Column(name = "client_operation_id", length = 120)
    private String clientOperationId;

    @Column(name = "start_offset", nullable = false)
    private long startOffset;

    @Column(name = "end_offset", nullable = false)
    private long endOffset;

    @Column(name = "highlight_text", columnDefinition = "text")
    private String highlightText;

    @Column(columnDefinition = "text")
    private String note;

    @Column(nullable = false, length = 20)
    private String color = "#FFEB3B";

    @Column(name = "created_at", nullable = false)
    private Instant createdAt;

    @Column(name = "updated_at", nullable = false)
    private Instant updatedAt;

    @Version
    @Column(nullable = false)
    private long version;

    @PrePersist
    void fillCreatedFields() {
        if (id == null) {
            id = UUID.randomUUID();
        }
        Instant now = Instant.now();
        if (createdAt == null) {
            createdAt = now;
        }
        if (updatedAt == null) {
            updatedAt = now;
        }
    }

    @PreUpdate
    void fillUpdatedAt() {
        updatedAt = Instant.now();
    }
}

File: frontend/lib/features/reader/domain/reader_annotation.dart
class ReaderAnnotation {
  const ReaderAnnotation({
    required this.id,
    required this.readerItemId,
    required this.startOffset,
    required this.endOffset,
    required this.color,
    required this.createdAt,
    this.chapterId,
    this.highlightText,
    this.note,
  });

  factory ReaderAnnotation.fromJson(Map<String, dynamic> json) {
    return ReaderAnnotation(
      id: json['id']?.toString() ?? '',
      readerItemId: json['readerItemId']?.toString() ?? '',
      startOffset: _asInt(json['startOffset']),
      endOffset: _asInt(json['endOffset']),
      chapterId: json['chapterId']?.toString(),
      highlightText: json['highlightText']?.toString(),
      note: json['note']?.toString(),
      color: json['color']?.toString() ?? '#FFEB3B',
      createdAt: _parseDateTime(json['createdAt']),
    );
  }

  final String id;
  final String readerItemId;
  final int startOffset;
  final int endOffset;

  /// 章节 ID（可选，用于按章节过滤批注）
  final String? chapterId;
  final String? highlightText;
  final String? note;
  final String color;
  final DateTime? createdAt;
}

int _asInt(dynamic value) => switch (value) {
  int() => value,
  num() => value.toInt(),
  _ => int.tryParse(value?.toString() ?? '') ?? 0,
};

DateTime? _parseDateTime(dynamic value) {
  if (value == null || value.toString().isEmpty) return null;
  return DateTime.tryParse(value.toString().replaceFirst(' ', 'T'))?.toLocal();
}
